stylix/testbed.nix

```
@@ -0,0 +1,94 @@
+{ pkgs, inputs, lib, ... }:
+
+let
+  username = "guest";
+
+  commonModule = { config, ... }: {
+    users.users.${username} = {
+      description = "Guest";
+      hashedPassword = "";
+      isNormalUser = true;
+    };
+
+    # The state version can safely track the latest release because the disk
+    # image is ephermal.
+    system.stateVersion = config.system.nixos.release;
+    home-manager.users.${username}.home.stateVersion = config.system.nixos.release;
+
+    virtualisation.vmVariant.virtualisation = {
+      # This is a maximum limit; the VM should still work if the host has fewer cores.
+      cores = 4;
+      memorySize = 2048;
+    };
+  };
+
+  autoload = builtins.concatLists
+    (lib.mapAttrsToList
+      (name: _:
+        let testbed = {
+          inherit name;
+          module = "${../modules}/${name}/testbed.nix";
+        };
+        in
+          lib.optional (builtins.pathExists testbed.module) testbed
+      )
+      (builtins.readDir ../modules));
+
+  makeTestbed =
+    testbed: stylix:
+    let
+      name = "testbed-${testbed.name}-${stylix.polarity}";
+
+      system = lib.nixosSystem {
+        inherit (pkgs) system;
+
+        modules = [
+          commonModule
+          inputs.self.nixosModules.stylix
+          inputs.home-manager.nixosModules.home-manager
+          testbed.module
+
+          {
+            inherit stylix;
+            system.name = name;
+          }
+        ];
+      };
+
+      script = pkgs.writeShellApplication {
+        inherit name;
+        text = ''
+          # We create a temporary directory rather than a temporary file, since
+          # temporary files are created empty and are not valid disk images.
+          directory="$(mktemp --directory)"
+
+          clean() {
+            if rm --recursive "$directory"; then
+              echo 'Virtualisation disk image removed.'
+            fi
+          }
+          trap clean EXIT
+
+          NIX_DISK_IMAGE="$directory/nixos.qcow2" \
+            ${lib.getExe system.config.system.build.vm}
```

I performed the following nitpicks:

- Move the function declaration to the top
- Replace `echo` with `printf '%s\n'`
- Rename the `clean` function to `cleanup`

```suggestion
          cleanup() {
            if rm --recursive "$directory"; then
              printf '%s\n' 'Virtualisation disk image removed.'
            fi
          }

          # We create a temporary directory rather than a temporary file, since
          # temporary files are created empty and are not valid disk images.
          directory="$(mktemp --directory)"
          trap cleanup EXIT

          NIX_DISK_IMAGE="$directory/nixos.qcow2" \
            ${lib.getExe system.config.system.build.vm}
```

Feel free to ignore them, if you prefer.